src/common/static_ptr.h

```diff
@@ -0,0 +1,409 @@
+  // incompatible. Also never try to eat anything bigger than you are.
+  //
+  template<typename T, std::size_t S>
+  constexpr static int create_ward() noexcept {
+    static_assert(std::is_void<Base>{} ||
+                  std::is_base_of<Base, decay_t<T>>{},
+                  "Value to store must be a derivative of the base.");
+    static_assert(S <= Size, "Value too large.");
+    static_assert(std::is_void<Base>{} || !std::is_const<Base>{} ||
+                  std::is_const<T>{},
+                  "Cannot assign const pointer to non-const pointer.");
+    return 0;
+  }
+  // Here we can store anything that has the same signature, which is
+  // relevant to the multiple-versions for move/copy support that I
+  // mentioned above.
+  //
+  size_t (*operate)(_mem::op, void*, void*) noexcept;
+
+  // This is mutable so that get and the dereference operators can be
+  // const. Since we're modeling a pointer, we should preserve the
+  // difference in semantics between a pointer-to-const and a const
+  // pointer.
+  //
+  mutable typename std::aligned_storage<Size>::type buf;
+
+public:
+  using element_type = Base;
+  using pointer = Base*;
+
+  // Empty
+  static_ptr() noexcept : operate(nullptr) {}
+  static_ptr(std::nullptr_t) noexcept : operate(nullptr) {}
+  static_ptr& operator =(std::nullptr_t) noexcept {
+    reset();
+    return *this;
+  }
+  ~static_ptr() noexcept {
+    reset();
+  }
+
+  // Since other pointer-ish types have it
+  void reset() noexcept {
+    if (operate) {
+      operate(_mem::op::destroy, &buf, nullptr);
+      operate = nullptr;
+    }
+  }
+
+  // Set from another static pointer.
+  //
+  template<typename U, std::size_t S>
+  static_ptr(const static_ptr<U, S>& rhs)
+    noexcept(std::is_nothrow_copy_constructible<U>{}) : operate(rhs.operate) {
+    create_ward<U, S>();
+    if (operate) {
+      operate(_mem::op::copy, &rhs.buf, &buf);
+    }
+  }
+  template<typename U, std::size_t S>
+  static_ptr(static_ptr<U, S>&& rhs)
+    noexcept(std::is_nothrow_move_constructible<U>{}) : operate(rhs.operate) {
+    create_ward<U, S>();
+    if (operate) {
+      operate(_mem::op::move, &rhs.buf, &buf);
+    }
+  }
+
+  template<typename U, std::size_t S>
+  static_ptr& operator =(const static_ptr<U, S>& rhs)
+    noexcept(std::is_nothrow_copy_constructible<U>{}) {
+    create_ward<U, S>();
+    reset();
+    if (rhs) {
+      operate = rhs.operate;
+      operate(_mem::op::copy, const_cast<void*>(static_cast<const void*>(&rhs.buf)), &buf);
+    }
+    return *this;
+  }
+  template<typename U, std::size_t S>
+  static_ptr& operator =(static_ptr<U, S>&& rhs)
+    noexcept(std::is_nothrow_move_constructible<U>{}) {
+    create_ward<U, S>();
+    reset();
+    if (rhs) {
+      operate = rhs.operate;
+      operate(_mem::op::copy, const_cast<void*>(static_cast<const void*>(&rhs.buf)), &buf);
+    }
+    operate = rhs.operate;
+    operate(_mem::op::move, &rhs.buf, &buf);
```

Comment:
this does both copy and move. i'm guessing you just want the move inside the if block